### PR TITLE
Allow directly reading web pages, even when SERP not enabled

### PR DESCRIPTION
**Pull request by debanjum**
### Overview
Khoj can now read website directly without needing to go through the search step first.

### Details
9974d031 Parallelize simple webpage read and extractor
81faa026 Rename extract_content online results field to web pages
50d77906 Tweak prompts to extract information from webpages, online results
83fc7d0d Test select webpage as data source and extract web urls chat actors

384db4ce Render webpage read in chat response references on Web, Desktop apps
0f9c7ede Pass multiple webpages with their urls in online results context

28d1afb1 Support webpage command in chat API
f8c23caa Add webpage chat command for read web pages requested by user
6118d1ff Create chat actor for directly reading webpages based on user message

### CODE REVIEW

File: src/khoj/routers/helpers.py [hunk 1]
```diff
@@ -229,6 +230,35 @@ async def aget_relevant_output_modes(query: str, conversation_history: dict):
...
+
+    utc_date = datetime.utcnow().strftime("%Y-%m-%d")
+    online_queries_prompt = prompts.infer_webpages_to_read.format(
+        current_date=utc_date,
+        query=q,
+        chat_history=chat_history,
+        location=location,
+    )
+
+    response = await send_message_to_model_wrapper(online_queries_prompt, response_type="json_object")
```
Comment (sabaimran) [inline src/khoj/routers/helpers.py:248]: Could we put timer context's around these calls or add it to the `send_message_to_model_wrapper` method? It'll be really useful to assess bottlenecks in our response time for later improvements.

File: src/khoj/utils/helpers.py [hunk 2]
```diff
@@ -287,6 +290,7 @@ class ConversationCommand(str, Enum):
     ConversationCommand.General: "Use this when you can answer the question without any outside information or personal knowledge",
     ConversationCommand.Notes: "To search the user's personal knowledge base. Especially helpful if the question expects context from the user's notes or documents.",
     ConversationCommand.Online: "To search for the latest, up-to-date information from the internet. Note: **Questions about Khoj should always use this data source**",
+    ConversationCommand.Webpage: "To use if the user has directly provided the webpage urls or you are certain of the webpage urls to read.",
```
Comment (sabaimran) [inline src/khoj/utils/helpers.py:293]: Good to unify the language here. I think `ConversationCommand.General` should be updated to match the `To <verb> if <blah>` pattern.

File: src/khoj/processor/tools/online_search.py [hunk 3]
```diff
@@ -93,19 +95,35 @@ def search_with_google(subquery: str):
...
+    logger.info(f"Inferring web pages to read")
+    urls = await infer_webpage_urls(query, conversation_history, location)
+
+    logger.info(f"Reading web pages at: {urls}")
+    tasks = [read_webpage_and_extract_content(query, url) for url in urls]
+    results = await asyncio.gather(*tasks)
+
+    response: Dict[str, Dict] = defaultdict(dict)
+    response[query]["webpages"] = [
+        {"link": url, "snippet": web_extract} for _, web_extract, url in results if web_extract is not None
```
Comment (sabaimran) [inline src/khoj/processor/tools/online_search.py:109]: Should we include the subquery as well?

File: src/khoj/processor/tools/online_search.py [hunk 4]
```diff
@@ -93,19 +95,35 @@ def search_with_google(subquery: str):
...
+    return response
+
+
+async def read_webpage_and_extract_content(subquery: str, url: str) -> Tuple[str, Union[None, str], str]:
     try:
         with timer(f"Reading web page at '{url}' took", logger):
-            content = await read_webpage_with_olostep(url) if OLOSTEP_API_KEY else await read_webpage(url)
+            content = await read_webpage_with_olostep(url) if OLOSTEP_API_KEY else await read_webpage_at_url(url)
         with timer(f"Extracting relevant information from web page at '{url}' took", logger):
             extracted_info = await extract_relevant_info(subquery, content)
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/processor/tools/online_search.py:119]: Yeah, makes sense as a future optimization. Also being able to prioritize items to remove from the most recent message context to fit within token limit would be great to have

File: src/khoj/processor/conversation/prompts.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
 
 - You *CAN REMEMBER ALL NOTES and PERSONAL INFORMATION FOREVER* that the user ever shares with you.
 - Users can share files and other information with you using the Khoj Desktop, Obsidian or Emacs app. They can also drag and drop their files into the chat window.
-- You can generate images, look-up information from the internet, and answer questions based on the user's notes.
+- You *CAN* generate images, look-up real-time information from the internet, and answer questions based on the user's notes.
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/processor/conversation/prompts.py:13]: In my limited testing adding real-time helped with the responses, so not changing it for now but more testing is required

**Pull request merged**